### PR TITLE
More precise signatures in concurrent.futures

### PR DESCRIPTION
**Pull request by pkch**
A few changes to the signatures in `concurrent.futures`.

The first two should reduce false positives:

- `Future.exception()` and `Future.set_exception()` should be ok with `None`
- `Executor.map()` (and subclasses) return `Iterator` not just `Iterable`

The last one is going in the other direction (increases the strictness of type checking):

- `Executor.map()` iterables argument can be typed as `Iterable[Any]` instead of `Any`

I don't believe it would cause false positives, but if I'm missing some scenario I can remove that last change.

### CODE REVIEW

**Comment by gvanrossum** [comment]
Hm, this makes me curious. Methods that are defined identically in the Executor base class should not be repeated in TheadPoolExecutor and ProcessPoolExecutor at all, right?

**Comment by pkch** [comment]
Right, fixed.

**Comment by JelleZijlstra** [comment]
Do any of these enhancements also need to be applied to the Python 2 backport in third_party/2/concurrent/futures?

**Comment by pkch** [comment]
Tried to fix that, assuming there are no python 2 - specific typing issues that I'm not aware of.

File: third_party/2/concurrent/futures/__init__.pyi [hunk 1]
```diff
@@ -15,27 +15,27 @@ class Future(Generic[_T]):
     def running(self) -> bool: ...
     def done(self) -> bool: ...
     def result(self, timeout: Optional[float] = ...) -> _T: ...
-    def exception(self, timeout: Optional[float] = ...) -> Any: ...
+    def exception(self, timeout: Optional[float] = ...) -> Optional[BaseException]: ...
```
Comment (JelleZijlstra) [inline third_party/2/concurrent/futures/__init__.pyi]: Actually the Any here is intentional; it was added to support old-style classes (#946).

File: third_party/2/concurrent/futures/__init__.pyi [hunk 2]
```diff
@@ -15,27 +15,27 @@ class Future(Generic[_T]):
...
     def submit(self, fn: Callable[..., _T], *args: Any, **kwargs: Any) -> Future[_T]: ...
-    def map(self, func: Callable[..., _T], *iterables: Any, timeout: Optional[float] = ...) -> Iterable[_T]: ...
+    def map(self, func: Callable[..., _T], *iterables: Iterable[Any], timeout: Optional[float] = ...) -> Iterator[_T]: ...
     def shutdown(self, wait: bool = ...) -> None: ...
     def __enter__(self) -> Executor: ...
     def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> bool: ...
 
 class ThreadPoolExecutor(Executor):
-    def __init__(self, max_workers: int) -> None: ...
+    def __init__(self, max_workers: Optional[int]) -> None: ...
```
Comment (JelleZijlstra) [inline third_party/2/concurrent/futures/__init__.pyi]: This should also have `= ...`.

**Pull request merged**